src/hotspot/share/gc/shared/barrierSetRuntime.cpp

```diff
@@ -28,12 +28,21 @@
 #include "runtime/interfaceSupport.inline.hpp"
 #include "utilities/macros.hpp"
 
+
 JRT_LEAF(void, BarrierSetRuntime::value_copy(void* src, void* dst, InlineKlass* md))
   assert(md->is_inline_klass(), "invariant");
-  HeapAccess<>::value_copy(src, dst, md);
+  HeapAccess<>::value_copy(src, dst, md, LayoutKind::PAYLOAD);  // FIXME Hard coded value for the transition
+JRT_END
+
+JRT_LEAF(void, BarrierSetRuntime::value_copy2(void* src, void* dst, InlineLayoutInfo* li))
+  HeapAccess<>::value_copy(src, dst, li->klass(), li->kind());
```

Comment:
This is temporary. The original value_copy() method is still there because of arrays. Once arrays have been updated to support the new flat layouts, value_copy() will be removed, and value_copy2() will be renamed value_copy().